src/linesearch/armijo_goldstein.jl

```diff
@@ -0,0 +1,150 @@
+export armijo_goldstein
+
+"""
+    t, ht, nbk, nbG = armijo_goldstein(h, h₀, slope)
+
+Performs a line search from `x` along the direction `d` as defined by the
+`LineModel` ``h(t) = f(x + t d)``, where
+`h₀ = h(0) = f(x)`, `slope = h'(0) = ∇f(x)ᵀd`.
+The steplength is chosen trying to satisfy the Armijo and Goldstein conditions. The Armijo condition is
+```math
+h(t) ≤ h₀ + τ₀ t h'(0)
+```
+and the Goldstein condition is
+```math
+h(t) ≥ h₀ + τ₁ t h'(0).
+```
+with `1 > τ₁ > τ₀ > 0`.
+
+# Arguments
+
+- `h::LineModel{T, S, M}`: 1-D model along the search direction `d`, ``h(t) = f(x + t d)``
+- `h₀::T`: value of `h` at `t=0`
+- `slope`: dot product of the gradient and the search direction, ``∇f(x_k)ᵀd``
+The keyword arguments may include
+- `t::T = one(T)`: starting steplength (default `1`);
+- `τ₀::T = T(eps(T)^(1/4))`: slope factor in the Armijo condition. It should satisfy `1 > τ₁ > τ₀ > 0`;
+- `τ₁::T = min(T(1)-eps(T),T(0.9999))`: slope factor in the Goldstein condition. It should satisfy `1 > τ₁ > τ₀ > 0`;
+- `γ₀::T = T(1 / 2)`: backtracking step length mutliplicative factor (0<γ₀<1)
+- `γ₁::T = T(2)`: look-aheqd step length mutliplicative factor (1<γ₁)
+- `bk_max`: maximum number of backtracks (default `10`);
+- `bG_max`: maximum number of increases (default `10`);
+- `verbose`: whether to print information (default `false`).
+
+# Outputs
+
+- `t::T`: the step length;
```

Comment:
Yes, it can happen if the backtrack budget is consumed before the bisection method produces a step length that satisfies Armijo.
I think I would worth leaving a budget but keeping track of the last computed step length that satisfies Armijo.